models/glif_psc.cpp

```diff
@@ -0,0 +1,578 @@
+
+  S_.y1_.resize( P_.n_receptors_() );
+  S_.y2_.resize( P_.n_receptors_() );
+  V_.PSCInitialValues_.resize( P_.n_receptors_() );
+
+  B_.spikes_.resize( P_.n_receptors_() );
+
+  double Tau_ = P_.C_m_ / P_.G_; // in ms
+  V_.P33_ = std::exp( -h / Tau_ );
+  V_.P30_ = 1 / P_.C_m_ * ( 1 - V_.P33_ ) * Tau_;
+
+  for ( size_t i = 0; i < P_.n_receptors_(); i++ )
+  {
+    // these P are independent
+    V_.P11_[ i ] = V_.P22_[ i ] = std::exp( -h / P_.tau_syn_[ i ] );
+
+    V_.P21_[ i ] = h * V_.P11_[ i ];
+
+    // these are determined according to a numeric stability criterion
+    // input time parameter shall be in ms, capacity in pF
+    V_.P31_[ i ] = propagator_31( P_.tau_syn_[ i ], Tau_, P_.C_m_, h );
+    V_.P32_[ i ] = propagator_32( P_.tau_syn_[ i ], Tau_, P_.C_m_, h );
+
+    V_.PSCInitialValues_[ i ] = 1.0 * numerics::e / P_.tau_syn_[ i ];
+    B_.spikes_[ i ].resize();
+  }
+}
+
+/* ----------------------------------------------------------------
+ * Update and spike handling functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::glif_psc::update( Time const& origin, const long from, const long to )
+{
+  // glif_lif_r_asc_a
+  const double dt = Time::get_resolution().get_ms();
+
+  // get model type
+  std::string model_str = P_.glif_model_;
+  std::transform( model_str.begin(), model_str.end(), model_str.begin(), ::tolower );
+  if ( nest::glif_psc::model_type_lu.find( model_str ) == nest::glif_psc::model_type_lu.end() )
+  {
+    throw BadProperty( "Bad glif model type string." );
+  }
+  long model_type = nest::glif_psc::model_type_lu[ model_str ];
+
+  double v_old = S_.U_;
+  double spike_component = 0.0;
+  double voltage_component = 0.0;
+  double th_old = S_.threshold_;
+
+  for ( long lag = from; lag < to; ++lag )
+  {
+
+    // update threshold via exact solution of dynamics of spike component of
+    // threshold
+    // exact solution of dynamics of spike component of threshold for glif2/4/5
+    // models with "R"
+    if ( model_type == 2 || model_type == 4 || model_type == 5 )
+    {
+      spike_component = V_.last_spike_ * std::exp( -P_.b_spike_ * dt );
+    }
+    V_.last_spike_ = spike_component;
+
+    if ( V_.t_ref_remaining_ > 0.0 )
+    {
+      // While neuron is in refractory period count-down in time steps (since dt
+      // may change while in refractory) while holding the voltage at last peak.
+      V_.t_ref_remaining_ -= dt;
+      if ( V_.t_ref_remaining_ <= 0.0 )
+      {
+        // Neuron has left refractory period, reset voltage and after-spike
+        // current
+
+        // Reset ASC_currentsfor glif3/4/5 models with "ASC"
+        if ( model_type == 3 || model_type == 4 || model_type == 5 )
+        {
+          for ( std::size_t a = 0; a < S_.ASCurrents_.size(); ++a )
+          {
+            S_.ASCurrents_[ a ] =
+              P_.asc_amps_[ a ] + S_.ASCurrents_[ a ] * P_.r_[ a ] * std::exp( -P_.k_[ a ] * V_.t_ref_total_ );
```

Comment:
Should comment here why you use `V_.t_ref_total_`, it is not obvious. I also don't quite manage to match this to Eqs 3 and 7 in [1], could you explain?

Reply:
This line is for ASC rest, i.e., Eq. 7 in [1] with f_j = exp ( - k_j * delta_t) (see texts right below Eq. (6) in the Supplementary Materials of [1]). delta_t is the total time of spike cut length (refractory period) as the after spike currents also decay during such a period. V_.t_ref_total_ in codes is delta_t.